### PR TITLE
Merging upstream changes.

### PR DESCRIPTION
**Pull request by jasonimercer**
These changes are the results of merging upstream navigation/indigo-devel with our navigation/indigo-devel and addressing merge conflicts.

The biggest change is the removal of the costmap's footprint layer.

CORE-1854

### CODE REVIEW

**Comment by jasonimercer** [comment]
@skaynama 
@paulbovbel 

Many changes are ros-lint related, seems upstream ran it since the last merge. I've been able to test this in simulation without any surprises.

**Comment by mikepurvis** [comment]
Looks like the formatting changes were all in 49c84c5a78722db0695dcbba21cedc3b2299e9b3. In the future it may make sense to cherry-pick a change like that in upfront, so that the "real" changes can be discussed without the distraction of the style changes.

File: costmap_2d/cfg/ObstaclePlugin.cfg [hunk 1]
```diff
@@ -5,6 +5,7 @@ from dynamic_reconfigure.parameter_generator_catkin import ParameterGenerator, b
 gen = ParameterGenerator()
 
 gen.add("enabled", bool_t, 0, "Whether to apply this plugin or not", True)
+gen.add("footprint_clearing_enabled", bool_t, 0, "Whether to clear the robot's footprint of lethal obstacles", True)
```
Comment (skaynama) [inline costmap_2d/cfg/ObstaclePlugin.cfg]: Why is this True by default?

Comment (jasonimercer) [inline costmap_2d/cfg/ObstaclePlugin.cfg]: Mainly to match upstream. Just finished a discussion that solidifies the case to have this turned off so I'll make that change.

File: costmap_2d/include/costmap_2d/inflation_layer.h [hunk 2]
```diff
@@ -111,15 +117,6 @@ class InflationLayer : public Layer
...
                            int min_i, int min_j, int max_i, int max_j);
 
-  /** @brief Apply inflation to master_grid 
-    * @param master_grid Costmap2D to operate on
-    * @param min_i Window bounds to apply inflation
-    * @param min_j Window bounds to apply inflation
-    * @param max_i Window bounds to apply inflation
-    * @param max_j Window bounds to apply inflation
-    */
-  virtual void updateCosts(Costmap2D &master_grid, int min_i, int min_j, int max_i, int max_j);
```
Comment (skaynama) [inline costmap_2d/include/costmap_2d/inflation_layer.h]: Wait, what happened to this method? They removed it or has it been relocated to somewhere else?

Comment (jasonimercer) [inline costmap_2d/include/costmap_2d/inflation_layer.h]: Moved to try to make it closer to upstream's structure. New line number if 102.

File: costmap_2d/plugins/inflation_layer.cpp [hunk 3]
```diff
@@ -190,14 +190,14 @@ void InflationLayer::updateCostsPQ(Costmap2D &master_grid, int min_i, int min_j,
...
-  //make sure the inflation queue is empty at the beginning of the cycle (should always be true)
+  // make sure the inflation queue is empty at the beginning of the cycle (should always be true)
   ROS_ASSERT_MSG(inflation_queue_.empty(), "The inflation queue must be empty at the beginning of inflation");
 
   unsigned char* master_array = master_grid.getCharMap();
   unsigned int size_x = master_grid.getSizeInCellsX(), size_y = master_grid.getSizeInCellsY();
 
   if (seen_ == NULL) {
-    ROS_ERROR("seen is NULL");
+    ROS_WARN("InflationLayer::updateCosts(): seen_ array is NULL");
```
Comment (skaynama) [inline costmap_2d/plugins/inflation_layer.cpp]: YES! Finally! (Not that the new log message is any more descriptive, haha)  @servos

File: costmap_2d/cfg/ObstaclePlugin.cfg (hunk 1 above)
Comment (skaynama) [inline costmap_2d/cfg/ObstaclePlugin.cfg]: Correct. This was the main reason we removed the footprint_layer plugin from the yaml files.

**Comment by skaynama** [comment]
Can't tell because of all the roslint changes, but has the API changed in any way?

**Comment by skaynama** [comment]
First pass LGTM.

**Comment by jasonimercer** [comment]
There are some changes but they are related to the footprint getting it's own class. The move_base changes are all additions.

**Comment by jasonimercer** [comment]
Since this is a rather large PR I'll see if we can get Davide or Kyle to do some sanity on it.

**Comment by skaynama** [comment]
Are the API changes documented anywhere yet?

**Comment by jasonimercer** [comment]
I really don't know how upstream publishes API changes.

**Comment by skaynama** [comment]
Looks fairly recent: http://docs.ros.org/indigo/api/costmap_2d/html/index.html

**Comment by jasonimercer** [comment]
@p6chen - when you get a chance, could you give this a once-over?

**Comment by p6chen** [comment]
Overall it looks ok to me. But given the number of changes in the PR, this probably needs to be run through tests to build confidence on it before merging.

**Comment by jasonimercer** [comment]
I agree 100%.

**Comment by paulbovbel** [comment]
generally LGTM

**Comment by jasonimercer** [comment]
Hi folks, I'm making a temporary branch of this with the default footprint clearing mechanism set to true to deal with lidar noise. The branch is CORE-1854-Clear-Footprint. This will allow testing on the robot even if the lidar is acting up.

**Comment by servos** [comment]
Isn't footprint_clearing just a param which we could/should be setting in the *_navigation file?

**Comment by skaynama** [comment]
Isn't it a dynamic reconfigure? Davide could just set it to true and run some tests, before we commit to any changing of the default value.

**Comment by jasonimercer** [comment]
At present it's set in navigation/costmap_2d/cfg/ObstaclePlugin.cfg and controlled via dynamic reconfigure

**Comment by jasonimercer** [comment]
I'm not advocating for change in this branch, I just wanted to give him a convenient way to test these changes with footprint clearing turned on. He could have just used this branch and made the change whenever move_base restarted, or edited the file by hand, but having the parameter set for him saves him a step.

**Pull request merged**